# CCBD SP26 – Variant 2: Partitioning Strategies
## Analysis Notebook

This notebook loads the benchmark results from `results.csv` and produces **interactive plots** with filters.
Use the dropdowns, toggles and checkboxes below each section to explore results across layouts and dataset sizes.

**Metrics analysed:**
- Object count and storage size per layout
- S3 listing time
- Upload and download throughput
- Selective query runtime
- Broad query runtime

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import ipywidgets as widgets
from IPython.display import display
import numpy as np

df = pd.read_csv("results.csv")

SIZES        = ["S", "M", "L"]
LAYOUT_ORDER = ["flat", "by_date", "by_region"]
COLORS       = {"flat": "#4C72B0", "by_date": "#DD8452", "by_region": "#55A868"}

print("Results loaded successfully.")
df

Results loaded successfully.


,size,layout,local_listing_time_s,local_file_count,local_total_size_gb,s3_listing_time_s,s3_object_count,s3_total_size_gb,upload_throughput_mbs,download_throughput_mbs,selective_time_s,selective_rows,broad_time_s,broad_rows
0,S,flat,0.0003,1,0.0737,0.1266,1,0.0737,179.34,468.44,0.1747,64587,0.1034,5000000
1,S,by_date,0.0158,157,0.1060,0.1103,157,0.1060,16.05,30.00,0.0588,96919,0.4787,5000000
2,S,by_region,0.0013,25,0.1029,0.0248,25,0.1029,40.17,168.74,0.0456,96919,0.1270,5000000
3,M,flat,0.0003,1,0.3686,0.0059,1,0.3686,261.08,408.89,1.2390,322200,0.8926,25000000
4,M,by_date,0.0698,784,0.5298,5.3419,784,0.5298,17.94,31.81,0.4943,483372,4.1051,25000000
5,M,by_region,0.0046,125,0.5145,0.1100,125,0.5145,38.91,151.32,0.3569,483372,1.1172,25000000
6,L,flat,0.0004,1,1.4745,0.0078,1,1.4745,246.18,425.03,5.1213,1291441,3.9777,100000000
7,L,by_date,0.2159,3136,2.1193,33.0353,3136,2.1193,8.27,34.06,3.5099,1936418,27.7179,100000000
8,L,by_region,0.0382,500,2.0580,0.2651,500,2.0580,57.28,88.49,1.6342,1936418,4.8968,100000000


## 1. Object Count and Storage Size

Select a metric and a dataset size to explore how many files each layout produces and how much storage it uses.
Partitioned layouts generate significantly more files due to Hive directory structure,
and are ~40% larger than flat due to per-file Parquet metadata overhead.

In [3]:
metric_selector = widgets.Dropdown(
    options=[("Number of files", "local_file_count"), ("Storage size (GB)", "local_total_size_gb")],
    value="local_file_count",
    description="Metric:",
    style={"description_width": "initial"},
)
size_toggle_1 = widgets.ToggleButtons(
    options=["S", "M", "L", "All"],
    value="All",
    description="Dataset size:",
    style={"description_width": "initial"},
)

def plot_storage(metric, size_filter):
    sizes = SIZES if size_filter == "All" else [size_filter]
    fig, ax = plt.subplots(figsize=(7, 4))
    x = np.arange(len(sizes))
    width = 0.25
    for i, layout in enumerate(LAYOUT_ORDER):
        subset = df[df["layout"] == layout].set_index("size").reindex(sizes)
        bars = ax.bar(x + i * width, subset[metric], width, label=layout,
                      color=COLORS[layout], edgecolor="white")
        for bar in bars:
            h = bar.get_height()
            ax.text(bar.get_x() + bar.get_width() / 2, h * 1.01,
                    f"{h:,.0f}" if metric == "local_file_count" else f"{h:.3f}",
                    ha="center", va="bottom", fontsize=8)
    label = "Number of files" if metric == "local_file_count" else "Size (GB)"
    ax.set_ylabel(label)
    ax.set_xticks(x + width)
    ax.set_xticklabels([f"Size {s}" for s in sizes])
    ax.set_title(f"{label} by layout")
    ax.legend(title="Layout")
    ax.spines[["top", "right"]].set_visible(False)
    plt.tight_layout()
    plt.show()

out1 = widgets.interactive_output(plot_storage, {"metric": metric_selector, "size_filter": size_toggle_1})
display(widgets.VBox([widgets.HBox([metric_selector, size_toggle_1]), out1]))

## 2. S3 Listing Time

Listing time measures how long it takes to enumerate all objects under a prefix via the S3 API.
With 3,136 objects at size L, by_date takes **33 seconds** to list — over 4,000x slower than flat.
Toggle between S3 and local listing to see the difference.

In [ ]:
layout_checks = widgets.SelectMultiple(
    options=LAYOUT_ORDER,
    value=LAYOUT_ORDER,
    description="Layouts:",
    style={"description_width": "initial"},
    rows=3,
)
listing_toggle = widgets.RadioButtons(
    options=[("S3 listing", "s3_listing_time_s"), ("Local listing", "local_listing_time_s")],
    value="s3_listing_time_s",
    description="Source:",
    style={"description_width": "initial"},
)

def plot_listing(layouts, metric):
    if not layouts:
        print("Select at least one layout.")
        return
    layouts = [l for l in LAYOUT_ORDER if l in layouts]
    fig, axes = plt.subplots(1, 3, figsize=(12, 4))
    title = "S3 listing time" if metric == "s3_listing_time_s" else "Local listing time"
    fig.suptitle(f"{title} (seconds) — lower is better", fontsize=12)
    for i, size in enumerate(SIZES):
        subset = df[(df["size"] == size) & (df["layout"].isin(layouts))].set_index("layout").reindex(layouts)
        bars = axes[i].bar(subset.index, subset[metric],
                           color=[COLORS[l] for l in subset.index], edgecolor="white")
        for bar in bars:
            h = bar.get_height()
            axes[i].text(bar.get_x() + bar.get_width() / 2, h * 1.01,
                         f"{h:.3f}s", ha="center", va="bottom", fontsize=8)
        axes[i].set_title(f"Size {size}")
        axes[i].set_ylabel("Seconds" if i == 0 else "")
        axes[i].spines[["top", "right"]].set_visible(False)
    plt.tight_layout()
    plt.show()

out2 = widgets.interactive_output(plot_listing, {"layouts": layout_checks, "metric": listing_toggle})
display(widgets.VBox([widgets.HBox([layout_checks, listing_toggle]), out2]))

## 3. Upload and Download Throughput

Select which direction (upload or download) and which sizes to compare.
Throughput is measured on a single representative sample file per layout.
Flat achieves up to 30x higher upload throughput than by_date at size L.

In [5]:
direction = widgets.RadioButtons(
    options=[("Upload", "upload_throughput_mbs"), ("Download", "download_throughput_mbs")],
    value="upload_throughput_mbs",
    description="Direction:",
    style={"description_width": "initial"},
)
size_checks = widgets.SelectMultiple(
    options=SIZES,
    value=SIZES,
    description="Sizes:",
    rows=3,
    style={"description_width": "initial"},
)

def plot_throughput(metric, sizes):
    if not sizes:
        print("Select at least one size.")
        return
    sizes = [s for s in SIZES if s in sizes]
    fig, ax = plt.subplots(figsize=(7, 4))
    x = np.arange(len(sizes))
    width = 0.25
    for i, layout in enumerate(LAYOUT_ORDER):
        subset = df[df["layout"] == layout].set_index("size").reindex(sizes)
        bars = ax.bar(x + i * width, subset[metric], width, label=layout,
                      color=COLORS[layout], edgecolor="white")
        for bar in bars:
            h = bar.get_height()
            ax.text(bar.get_x() + bar.get_width() / 2, h * 1.01,
                    f"{h:.0f}", ha="center", va="bottom", fontsize=8)
    label = "Upload (MB/s)" if metric == "upload_throughput_mbs" else "Download (MB/s)"
    ax.set_ylabel(label)
    ax.set_xticks(x + width)
    ax.set_xticklabels([f"Size {s}" for s in sizes])
    ax.set_title(f"{label} by layout")
    ax.legend(title="Layout")
    ax.spines[["top", "right"]].set_visible(False)
    plt.tight_layout()
    plt.show()

out3 = widgets.interactive_output(plot_throughput, {"metric": direction, "sizes": size_checks})
display(widgets.VBox([widgets.HBox([direction, size_checks]), out3]))

## 4. Query Runtime Comparison

Select the query type and dataset size to compare layouts.
- **Selective query**: filters on `region = "Zurich"` and a 3-day date range — partition pruning applies.
- **Broad query**: full scan, no filter — all files must be opened, file-count overhead dominates.

In [6]:
query_type = widgets.RadioButtons(
    options=[("Selective query", "selective_time_s"), ("Broad query", "broad_time_s")],
    value="selective_time_s",
    description="Query type:",
    style={"description_width": "initial"},
)
size_toggle_4 = widgets.ToggleButtons(
    options=["S", "M", "L", "All"],
    value="All",
    description="Dataset size:",
    style={"description_width": "initial"},
)

def plot_query(metric, size_filter):
    sizes = SIZES if size_filter == "All" else [size_filter]
    ncols = len(sizes)
    fig, axes = plt.subplots(1, ncols, figsize=(4 * ncols, 4))
    if ncols == 1:
        axes = [axes]
    label = "Selective query time (s)" if metric == "selective_time_s" else "Broad query time (s)"
    fig.suptitle(f"{label} — lower is better", fontsize=12)
    for i, size in enumerate(sizes):
        subset = df[df["size"] == size].set_index("layout").reindex(LAYOUT_ORDER)
        bars = axes[i].bar(LAYOUT_ORDER, subset[metric],
                           color=[COLORS[l] for l in LAYOUT_ORDER], edgecolor="white")
        for bar in bars:
            h = bar.get_height()
            axes[i].text(bar.get_x() + bar.get_width() / 2, h * 1.01,
                         f"{h:.3f}s", ha="center", va="bottom", fontsize=8)
        axes[i].set_title(f"Size {size}")
        axes[i].set_ylabel("Seconds" if i == 0 else "")
        axes[i].spines[["top", "right"]].set_visible(False)
    plt.tight_layout()
    plt.show()

out4 = widgets.interactive_output(plot_query, {"metric": query_type, "size_filter": size_toggle_4})
display(widgets.VBox([widgets.HBox([query_type, size_toggle_4]), out4]))

## 5. Head-to-Head Layout Comparison

Select two layouts and a dataset size to compare them directly across all metrics at once.

In [7]:
layout_a = widgets.Dropdown(
    options=LAYOUT_ORDER, value="flat", description="Layout A:",
    style={"description_width": "initial"},
)
layout_b = widgets.Dropdown(
    options=LAYOUT_ORDER, value="by_region", description="Layout B:",
    style={"description_width": "initial"},
)
size_h2h = widgets.Dropdown(
    options=SIZES, value="L", description="Dataset size:",
    style={"description_width": "initial"},
)

METRICS_H2H = [
    ("s3_listing_time_s",       "S3 listing (s)"),
    ("upload_throughput_mbs",   "Upload (MB/s)"),
    ("download_throughput_mbs", "Download (MB/s)"),
    ("selective_time_s",        "Selective (s)"),
    ("broad_time_s",            "Broad (s)"),
]

def plot_h2h(la, lb, size):
    if la == lb:
        print("Select two different layouts.")
        return
    row_a = df[(df["layout"] == la) & (df["size"] == size)].iloc[0]
    row_b = df[(df["layout"] == lb) & (df["size"] == size)].iloc[0]
    keys   = [m[0] for m in METRICS_H2H]
    labels = [m[1] for m in METRICS_H2H]
    vals_a = [row_a[k] for k in keys]
    vals_b = [row_b[k] for k in keys]
    x = np.arange(len(keys))
    width = 0.35
    fig, ax = plt.subplots(figsize=(10, 4))
    bars_a = ax.bar(x - width/2, vals_a, width, label=la, color=COLORS[la], edgecolor="white")
    bars_b = ax.bar(x + width/2, vals_b, width, label=lb, color=COLORS[lb], edgecolor="white")
    for bar in list(bars_a) + list(bars_b):
        h = bar.get_height()
        ax.text(bar.get_x() + bar.get_width() / 2, h * 1.01,
                f"{h:.2f}", ha="center", va="bottom", fontsize=8)
    ax.set_xticks(x)
    ax.set_xticklabels(labels)
    ax.set_title(f"Head-to-head: {la} vs {lb} — Size {size}")
    ax.legend(title="Layout")
    ax.spines[["top", "right"]].set_visible(False)
    plt.tight_layout()
    plt.show()

out5 = widgets.interactive_output(plot_h2h, {"la": layout_a, "lb": layout_b, "size": size_h2h})
display(widgets.VBox([widgets.HBox([layout_a, layout_b, size_h2h]), out5]))

## 6. Analysis Summary

In [8]:
print("\n=== ANALYSIS SUMMARY ===\n")

max_files = df.loc[df["local_file_count"].idxmax()]
print("1. Partitioned layouts create significantly more files than flat.")
print(f"   {max_files['size']}/by_date generates {int(max_files['local_file_count'])} files vs 1 for flat.")

worst_s3 = df.loc[df["s3_listing_time_s"].idxmax()]
worst_local = df.loc[
    (df["size"] == worst_s3["size"]) & (df["layout"] == worst_s3["layout"]),
    "local_listing_time_s"
].values[0]
ratio_listing = worst_s3["s3_listing_time_s"] / worst_local
print("\n2. S3 listing time scales badly with file count.")
print(f"   {worst_s3['size']}/{worst_s3['layout']}: {worst_s3['s3_listing_time_s']:.1f}s on S3 "
      f"vs {worst_local:.2f}s local (x{ratio_listing:.0f} overhead).")

l_sel = df[df["size"] == "L"].set_index("layout")["selective_time_s"]
ratio_sel = l_sel["flat"] / l_sel["by_region"]
print("\n3. Selective queries benefit most from by_region partitioning.")
print(f"   L: flat={l_sel['flat']:.2f}s, by_date={l_sel['by_date']:.2f}s, "
      f"by_region={l_sel['by_region']:.2f}s (x{ratio_sel:.1f} faster than flat).")

l_broad = df[df["size"] == "L"].set_index("layout")["broad_time_s"]
print("\n4. Broad queries favor flat layout (single sequential file read).")
print(f"   L: flat={l_broad['flat']:.2f}s, by_region={l_broad['by_region']:.2f}s, "
      f"by_date={l_broad['by_date']:.2f}s.")

l_up = df[df["size"] == "L"].set_index("layout")["upload_throughput_mbs"]
ratio_up = l_up["flat"] / l_up["by_date"]
print("\n5. Transfer throughput is much higher for flat (fewer, larger files).")
print(f"   L upload: flat={l_up['flat']:.0f} MB/s vs by_date={l_up['by_date']:.1f} MB/s "
      f"(x{ratio_up:.0f} difference).")

print("\n=== RECOMMENDATION ===")
print("- Use by_region for selective/analytical queries on a specific region.")
print("- Use flat for broad scans or when transfer speed matters.")
print("- Avoid by_date at large scale: listing overhead on S3 becomes prohibitive.")


=== ANALYSIS SUMMARY ===

1. Partitioned layouts create significantly more files than flat.
   L/by_date generates 3136 files vs 1 for flat.

2. S3 listing time scales badly with file count.
   L/by_date: 33.0s on S3 vs 0.22s local (x153 overhead).

3. Selective queries benefit most from by_region partitioning.
   L: flat=5.12s, by_date=3.51s, by_region=1.63s (x3.1 faster than flat).

4. Broad queries favor flat layout (single sequential file read).
   L: flat=3.98s, by_region=4.90s, by_date=27.72s.

5. Transfer throughput is much higher for flat (fewer, larger files).
   L upload: flat=246 MB/s vs by_date=8.3 MB/s (x30 difference).

=== RECOMMENDATION ===
- Use by_region for selective/analytical queries on a specific region.
- Use flat for broad scans or when transfer speed matters.
- Avoid by_date at large scale: listing overhead on S3 becomes prohibitive.
